📝 Add schema view and drift comparison to API endpoints docs (#37785)
DataDog/documentation: content/en/security/application_security/api_posture/api_inventory/api_endpoints.md

The {{< ui >}}Spans{{< /ui >}} data source shows real traffic and data exposure. Remediation should be performed in code, config, or access controls immediately.

> The actions you take depend on each of the attack surfaces:

Reviewer: ```suggestion
The actions you take depend on the attack surface:
```

Without explicit `codeLocations`, endpoints may not merge correctly with data from other sources.

## View and compare endpoint schemas

> API Posture builds an OpenAPI schema for each endpoint from the traffic it observes. This *inferred* schema describes what your API exposes in production: its paths, parameters, request and response bodies, and authentication. When your team also publishes a *declared* schema, an OpenAPI definition registered in the Datadog Software Catalog, you can compare the two to find where the running API has diverged from its documentation.

Reviewer: Italics are used to introduce the key term "inferred" (and "declared"). The word list recommends bold (used sparingly) for key terms on first introduction, not italics.

```suggestion
API Posture builds an OpenAPI schema for each endpoint from the traffic it observes. This **inferred** schema describes what your API exposes in production: its paths, parameters, request and response bodies, and authentication. When your team also publishes a **declared** schema, an OpenAPI definition registered in the Datadog Software Catalog, you can compare the two to find where the running API has diverged from its documentation.
```

### View an endpoint's schema

> 

Reviewer: Can we add a screenshot of what it looks like to "view an endpoint"? And perhaps a quick UI instructions like "In App & API Protection, click an endpoint to open a detailed side panel" or something along those lines.

Author: Added the screenshot in 5488ba24e2 (static/images/security/application_security/api/api_endpoint_definition_schema.png). While adding it, I also noticed the bullet labels didn't match the actual button text in the UI ("View Raw Schema" / "View Inferred Schemas"), so I aligned the doc wording to that.

> When viewing an endpoint, the **Definition** section displays the endpoint request parameters, request body and responses. Fields that contain sensitive data are marked with the type of sensitive data observed.

Author: Fixed in 001fb53405.

> When the endpoint is associated with an API in Datadog Software Catalog, the displayed information is based on the OpenAPI specification. Otherwise, the information is *inferred* from live traffic.

Reviewer: ```suggestion

When the endpoint is associated with an API in Datadog Software Catalog, the displayed information is based on the OpenAPI specification. Otherwise, the information is inferred from live traffic.
```

> - **View raw schema**: View the raw OpenAPI schema as YAML.

Reviewer: "Raw schema" here vs. "declared schema" used in the section heading and intro (lines 134, 146, 148) — these appear to refer to the same concept. Use one term consistently.

Author: Updated in c9a6c5f9d199728d60c1e2c3761d4b59be2f3e85

Author: Actually changed it back to "raw schema" as it's the label we see on the screenshot

> - **View raw schema**: View the raw OpenAPI schema as YAML.

Reviewer: Can we add a short lead-in before these bullets? I'm not sure what they're referring to -- are they sections in the side panel? Buttons?

Author: Fixed in c15b6d1393 — added a lead-in clarifying both actions live in the endpoint's Definition section.

> - **View inferred schema**: view the inferred schema as a preview or YAML. This inferred schema can be exported as an OpenAPI file in YAML or JSON.

Author: Fixed in 82d8f1e0ba.

> To reduce noise, the inferred schema only includes fields that are observed enough times, and automatically drops fields that haven't been seen recently. This keeps one-off traffic, such as a single malformed request or an attacker probing an endpoint with an unexpected field, from polluting the inferred schema and appearing as drift when compared against the declared schema.

Reviewer: "enough times" is pretty vague. Can we be more specific?

Author: Fixed in c15b6d1393 — specified the actual thresholds (observed at least 3 times, pruned after 7 days without being observed again).

> ### Compare *declared* and *inferred* schemas

Author: Fixed in 30a7566e46.

> ### Compare *declared* and *inferred* schemas

Reviewer: Italics in a heading is non-standard for this doc set — headings should be plain sentence-case text.

```suggestion
### Compare declared and inferred schemas
```

> To compare schemas, set up the following:

Author: Fixed in b7caabebcc.

- To view an inferred schema, [enable App and API Protection][9] on the service so endpoints are discovered from live traffic.
- To compare against a declared schema, register the API's OpenAPI definition in the Datadog Catalog. See [Create Entities][8].

> After both are in place, the differences appear directly on the endpoint's schema view, highlighted by severity:

Reviewer: ```suggestion
To compare inferred and declared schemas, you must:

- [Enable App and API Protection][9] on the service so endpoints are discovered from live traffic.
- Register the declared schema's OpenAPI definition in the Datadog Software Catalog. See [Create Entities][8].

The schema differences appear directly in the endpoint's schema view, highlighted by severity:
```

| Severity | Meaning |
|----------|---------|
| Breaking | The change likely breaks clients that rely on the declared contract, such as a field that became required or a parameter type change. |
| Warning | The change is drift worth reviewing, such as an undeclared field observed in traffic or a parameter that became optional. |

> | Info | The difference is low-risk, such as an endpoint that is declared but has no observed traffic. |

Reviewer: "low-risk" is a compound modifier used after the noun ("is low-risk"), which per style guide should generally not be hyphenated (not on the always-hyphenated exceptions list).

```suggestion
| Info | The difference is low risk, such as an endpoint that is declared but has no observed traffic. |
```

Differences can appear in the following areas of the schema:

- **Parameters**: A parameter added, removed, or changed from optional to required (or the reverse).
- **Request body**: A request body added, removed, or changed from optional to required (or the reverse).
- **Schema properties**: A property added, removed, changed from optional to required (or the reverse), or changed type, format, nullability, or enum values.

> - **Value constraints**: Changes to numeric or length limits (`minimum`, `maximum`, `minLength`, `maxLength`), patterns, or uniqueness constraints.

Reviewer: Making this bullet parallel with the rest.
```suggestion
- **Value constraints**: A numeric or length limit (`minimum`, `maximum`, `minLength`, `maxLength`), pattern, or uniqueness constraint changed. 
```

> - **Schema composition**: A mismatch in `oneOf` or `allOf` composition, or in a discriminator.

Reviewer: Making this parallel with the rest.

```suggestion
- **Schema composition**:  A mismatch introduced in `oneOf` or `allOf` composition, or in a discriminator.
```

- **Responses**: A status code, response header, or content type added or removed.

To reduce noise, some differences are excluded because they don't represent meaningful contract drift:

- Request header and cookie parameters, since they often carry values such as authentication tokens or session identifiers that aren't part of the API contract.
- Type changes on query parameters, since query parameters are always observed as strings in traffic, even when declared as another type, such as an integer or boolean.
- Removed status codes, since the inferred schema only includes status codes observed in traffic, so a declared status code that hasn't occurred yet during observation always appears as removed.

> - `anyOf` composition mismatches, since the declared and inferred schemas can use `anyOf` at different levels of the schema while remaining equivalent.

Reviewer: Updating format to match the list above and improve scannability

```suggestion
- **Request header and cookie parameters:** They often carry values such as authentication tokens or session identifiers that aren't part of the API contract.
- **Type changes on query parameters:** Query parameters are always observed as strings in traffic, even when declared as another type, such as an integer or Boolean.
- **Removed status codes:** The inferred schema only includes status codes observed in traffic, so a declared status code that hasn't occurred yet during observation always appears as removed.
- **`anyOf` composition mismatches:** The declared and inferred schemas can use `anyOf` at different levels of the schema while remaining equivalent.
```

### View an endpoint's schema

In [API Endpoints][1], click an endpoint to open its detail panel. The **Definition** section displays the endpoint's request parameters, request body, and responses. Fields that contain sensitive data are marked with the type of sensitive data observed.

{{< img src="/security/application_security/api/api_endpoint_definition_schema.png" alt="The Definition section of an endpoint's detail panel, showing its request parameters and the View Raw Schema and View Inferred Schemas buttons" style="width:100%;" >}}

When the endpoint is associated with an API in Datadog Software Catalog, the **Definition** section displays the declared OpenAPI specification. Otherwise, it displays the schema inferred from live traffic.

In the **Definition** section, you can:

- {{< ui >}}View Raw Schema{{< /ui >}}: View the displayed schema as raw YAML.
- {{< ui >}}View Inferred Schemas{{< /ui >}}: View the schema inferred from live traffic as a preview or YAML, even when a declared schema is available. The inferred schema can be exported as an OpenAPI file in YAML or JSON.

> To reduce noise, the inferred schema only includes fields observed at least three times, and drops fields that haven't been observed again within 7 days. This keeps one-off traffic, such as a single malformed request or an attacker probing an endpoint with an unexpected field, from polluting the inferred schema and appearing as drift when compared against the declared schema.

Reviewer: ```suggestion
To reduce noise, the inferred schema only includes fields observed at least three times, and drops fields that haven't been observed again within 7 days. This keeps one-off traffic, such as a single malformed request or an attacker probing an endpoint with an unexpected field, from polluting the inferred schema. Otherwise, it could appear as drift when compared against the declared schema.
```

The schema differences appear directly in the endpoint's schema view, highlighted by severity:

| Severity | Meaning |
|----------|---------|
| Breaking | The change likely breaks clients that rely on the declared contract, such as a field that became required or a parameter type change. |
| Warning | The change is drift worth reviewing, such as an undeclared field observed in traffic or a parameter that became optional. |
| Info | The difference is low risk, such as an endpoint that is declared but has no observed traffic. |

Differences can appear in the following areas of the schema:

- **Parameters**: A parameter added, removed, or changed from optional to required (or the reverse).
- **Request body**: A request body added, removed, or changed from optional to required (or the reverse).
- **Schema properties**: A property added, removed, changed from optional to required (or the reverse), or changed type, format, nullability, or enum values.
- **Value constraints**: A numeric or length limit (`minimum`, `maximum`, `minLength`, `maxLength`), pattern, or uniqueness constraint changed.

> - **Schema composition**:  A mismatch introduced in `oneOf` or `allOf` composition, or in a discriminator.

Reviewer: ```suggestion
- **Schema composition**: A mismatch introduced in `oneOf` or `allOf` composition, or in a discriminator.
```

Authentication is determined by:

- The presence of `Authorization`, `Token`, or `X-Api-Key` headers.
- The presence of a user ID within the trace (for example, the `@usr.id` APM attribute).

> - The endpoint returned a 401 or 403 status code.

Reviewer: Breaks parallel structure with sibling list items, which are all noun phrases ("The presence of...", "Custom... rules..."). Since this line was touched in the diff, consider aligning it while here.

```suggestion
- A 401 or 403 status code returned by the endpoint.
```
